### ISSUE TITLE
1.4 Beta1: variable not initialized in generated mapper in case of constructor mapping and mapper function

### ISSUE THREAD

**Issue by imeszaros** [mapstruct/mapstruct#2124]
- [x] Is this an issue (and hence not a question)? 

Mapper code:

```
    @Mappings(
            ...
            Mapping(target = "issueId", source = "commitComment.issueId", qualifiedBy = [ Mapped::class ]),
            ...)
    abstract fun clone(commitComment: CommitComment, forRepository: Repository, @Context cloneContext: CloneContext): CommitComment

    ...

    @Mapped
    fun mapIssueNumber(issueNumber: Int, @Context cloneContext: CloneContext) = cloneContext.issueNumberMapper(issueNumber)
```

Generated code:

```
    @Override
    public CommitComment clone(CommitComment commitComment, Repository forRepository, CloneContext cloneContext) {
        ...
        Integer issueId;
        ...
        if ( commitComment != null ) {
            ...
            if ( commitComment.getIssueId() != null ) {
                issueId = mapIssueNumber( commitComment.getIssueId().intValue(), cloneContext );
            }
            ...
        }
        else {
            ...
            issueId = null;
            ...
        }

        ...

        CommitComment commitComment1 = new CommitComment(..., issueId, ...);

        return commitComment1;
    }
```

Now the java compiler complains that the `issueId` (declared as `Int?` i.e. nullable) variable is not initialized, what could really happen if `commitComment.getIssueId()` returns `null`. Probably the generated code should initialize the variables at declaration (this way the else block could be omitted also?)?

Additionally, the mapper method `mapIssueNumber` is not picked up if the `source` attribute is omitted in the `Mapping` declaration. I guess it should work without it too?

**Comment by imeszaros** [comment]
Workaround found, add `defaultExpression = "java(null)"` to the `Mapping` definition.

**Comment by filiphr** [comment]
Thanks a lot for the feedback. It is extremely valuable to us so we can iron out the small details before the final release. Please let us know if you experience other issues as well.

I am 99% sure that this is fixed on master via da37d40. I am leaving this issue open so we can add a test case with your example.

**Comment by filiphr** [comment]
Regarding 

>Additionally, the mapper method `mapIssueNumber` is not picked up if the `source` attribute is omitted in the `Mapping` declaration. I guess it should work without it too?

This is an entirely different problem, the qualifier is ignored when name based mapping is done. Not sure whether we should take the selection (if it exists). Can you please create a separate issue for that? That has been like that before, so it is more of an enhancement.
